### PR TITLE
Support optional calls in no-unused-expressions

### PR DESCRIPTION
**Pull request by lehni**
This relates to the issue outlined here: https://github.com/babel/babel-eslint/issues/643

### CODE REVIEW

File: tests/rules/no-unused-expressions.js [hunk 1]
```diff
@@ -136,8 +136,9 @@ ruleTester.run("no-unused-expressions", rule, {
         },
 
         // Babel-specific test cases.
-      { code: "let a = do { foo; let b = 2; }", errors: [{ message: "Expected an assignment or function call and instead saw an expression.", type: "ExpressionStatement" }] },
+        { code: "let a = do { foo; let b = 2; }", errors: [{ message: "Expected an assignment or function call and instead saw an expression.", type: "ExpressionStatement" }] },
         { code: "let a = do { if (foo) { foo.bar } else { a; bar.foo } }", errors: [{ message: "Expected an assignment or function call and instead saw an expression.", type: "ExpressionStatement" }] },
+        { code: "function foo() { foo.bar?.(); }", errors: [] },
```
Comment (kaicataldo) [inline tests/rules/no-unused-expressions.js]: Shouldn't this be `foo.bar?.baz()`? What does this mean without a method name identifier?

Comment (kaicataldo) [reply tests/rules/no-unused-expressions.js]: Ah, also, this is valid, correct? If so, it should be in the `valid` section of this test suite.

Comment (ljharb) [reply tests/rules/no-unused-expressions.js]: This means, `foo.bar` is a normal GET, but `if `foo.bar` is nullish, it won't be invoked.

Comment (kaicataldo) [reply tests/rules/no-unused-expressions.js]: Cool, thanks for the explanation! Do you feel like this should be the default test case for this?

Comment (ljharb) [reply tests/rules/no-unused-expressions.js]: i don't know why it needs to be in a function; but anything that verifies that `foo.bar?.()` is the same re unused expressions as `foo.bar()` seems fine

Comment (lehni) [reply tests/rules/no-unused-expressions.js]: You're both right, thanks! Just to clarify: This test is specifically about optional calls, not optional chaining in general.

**Comment by lehni** [comment]
@kaicataldo I've addressed the things in the feedback now. Please take a look.

File: tests/rules/no-unused-expressions.js [hunk 2]
```diff
@@ -74,12 +74,13 @@ ruleTester.run("no-unused-expressions", rule, {
...
-      "let a = do { (foo + 1) }",
-      "let a = do { if (foo) { if (foo.bar) { foo.bar } } }",
-      "let a = do { if (foo) { if (foo.bar) { foo.bar } else if (foo.baz) { foo.baz } } }",
+        "let a = do { if (foo) { foo.bar } }",
+        "let a = do { foo }",
+        "let a = do { let b = 2; foo; }",
+        "let a = do { (foo + 1) }",
+        "let a = do { if (foo) { if (foo.bar) { foo.bar } } }",
+        "let a = do { if (foo) { if (foo.bar) { foo.bar } else if (foo.baz) { foo.baz } } }",
+        "foo.bar?.();",
```
Comment (lehni) [inline tests/rules/no-unused-expressions.js]: @kaicataldo these lines had wrong indentation, so I also fixed that along the way. I also noticed that the code uses semi-colons inconsistently, should probably be addressed for readability?

Comment (existentialism) [reply tests/rules/no-unused-expressions.js]: @lehni re: semicolons, I don't think it's a big deal, but fine if we make the babel-specific ones consistent. The others are just copied from ESLint directly.

Comment (lehni) [reply tests/rules/no-unused-expressions.js]: @existentialism ok, shall I do so in this PR?

Comment (lehni) [reply tests/rules/no-unused-expressions.js]: (the others don't have the same semi-colon style, btw)

**Comment by lehni** [comment]
@existentialism I've addressed all things now. Is this good to go?

**Comment by existentialism** [comment]
@lehni thanks!

**Comment by existentialism** [comment]
@lehni published as `5.2.0`!

**Pull request merged**